internal/security/secretstore/certs.go

```diff
@@ -0,0 +1,220 @@
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	req.Header.Set(VaultToken, t)
+	resp, err := cs.client.Do(req)
+	if err != nil {
+		e := fmt.Errorf("failed to retrieve the proxy cert on path %s with error %s", cs.certPath, err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+	defer resp.Body.Close()
+
+	cc := CertCollect{}
+
+	if resp.StatusCode == http.StatusNotFound {
+		e := fmt.Errorf("proxy cert pair NOT found in secret store @/%s, status: %s", cs.certPath, resp.Status)
+		LoggingClient.Info(e.Error())
+	} else if resp.StatusCode != http.StatusOK {
+		e := fmt.Errorf("failed to retrieve the proxy cert pair on path %s with error code %d", cs.certPath, resp.StatusCode)
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	if err = json.NewDecoder(resp.Body).Decode(&cc); err != nil {
+		e := fmt.Errorf("Error decoding json response when retrieving proxy cert pair: %s", err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, err
+	}
+
+	return &cc.Pair, nil
+}
+
+func (cs *Certs) AlreadyinStore() (bool, error) {
+	cp, err := cs.getCertPair()
+	if err != nil {
+		return false, err
+	}
+	if len(cp.Cert) > 0 && len(cp.Key) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (cs *Certs) getCertPair() (*CertPair, error) {
+	t, err := GetAccessToken(cs.tokenPath)
+	if err != nil {
+		return &CertPair{"", ""}, err
+	}
+	cp, err := cs.retrieve(t)
+	if err != nil {
+		return &CertPair{"", ""}, err
+	}
+	return cp, nil
+}
+
+func (cs *Certs) ReadFrom(certPath string, keyPath string) (*CertPair, error) {
+	certPEMBlock, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, err
+	}
+	cert := string(certPEMBlock)
+
+	keyPEMBlock, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		return nil, err
+	}
+	key := string(keyPEMBlock)
+
+	cp := CertPair{
+		Cert: cert,
+		Key:  key,
+	}
+
+	return &cp, nil
+}
+
+func GetAccessToken(filename string) (string, error) {
+	a := auth{}
+	raw, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return "", err
+	}
+	err = json.Unmarshal(raw, &a)
+	if err != nil {
+		return "", err
+	}
+
+	return a.Token, nil
+}
+
+func (cs *Certs) UploadToStore(cp *CertPair) error {
+	t, err := GetAccessToken(cs.tokenPath)
+	if err != nil {
+		return err
+	}
+
+	LoggingClient.Info("trying to upload the proxy cert pair into secret store")
+	jsonBytes, err := json.Marshal(cp)
+	body := bytes.NewBuffer(jsonBytes)
+
+	url, err := cs.certPathUrl()
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, url, body)
+	if err != nil {
+		e := fmt.Errorf("error creating http request: %v", err.Error())
+		LoggingClient.Error(e.Error())
+		return e
+	}
+
+	req.Header.Set(VaultToken, t)
+	resp, err := cs.client.Do(req)
+	if err != nil {
+		e := fmt.Sprintf("failed to upload the proxy cert pair on path %s with error %s", cs.certPath, err.Error())
+		LoggingClient.Error(e)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if !statusSuccess(resp.StatusCode) {
```

Comment:
couldn't there also be a 409 which means conflict (i.e. there's already one with the same name uploaded)?

Reply:
We check for it existing in line 77 of the main.go file

Reply:
We check to see if it's already uploaded before this happens in line 77 of main.go

Reply:
but there's a race condition there, where it didn't exist before and it exists now, so we should still handle that case here, but perhaps in this case we should error out if we see a 409 since someone interrupted our flow here

Reply:
Yeah, the error does seem appropriate here.  I will leave it as-is.